cpp/daal/src/algorithms/tsne/tsne_gradient_descent_impl.i

```diff
@@ -0,0 +1,1050 @@
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel2 = std::chrono::high_resolution_clock::now();
+        treeBuild += std::chrono::duration<double, std::milli>(kernel2 - kernel1).count();
+
+        status = summarizationKernelImpl<IdxType, DataType, cpu>(count, child, mass, posx, posy, nNodes, N, bottom);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel3 = std::chrono::high_resolution_clock::now();
+        summarization += std::chrono::duration<double, std::milli>(kernel3 - kernel2).count();
+
+        status = sortKernelImpl<IdxType, cpu>(sort, count, start, child, nNodes, N, bottom);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel4 = std::chrono::high_resolution_clock::now();
+        sorting += std::chrono::duration<double, std::milli>(kernel4 - kernel3).count();
+
+        status =
+            repulsionKernelImpl<IdxType, DataType, cpu>(theta, eps, sort, child, mass, posx, posy, repx, repy, zNorm, nNodes, N, radius, maxDepth);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel5 = std::chrono::high_resolution_clock::now();
+        repulsion += std::chrono::duration<double, std::milli>(kernel5 - kernel4).count();
+
+        if (((i + 1) % nIterCheck == 0) || (i == explorationIter - 1))
+        {
+            status = attractiveKernelImpl<true, IdxType, DataType, cpu>(val, col, row, posx, posy, attrx, attry, zNorm, divergence, nNodes, N, nnz,
+                                                                        nElements, exaggeration);
+        }
+        else
+        {
+            status = attractiveKernelImpl<false, IdxType, DataType, cpu>(val, col, row, posx, posy, attrx, attry, zNorm, divergence, nNodes, N, nnz,
+                                                                         nElements, exaggeration);
+        }
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel6 = std::chrono::high_resolution_clock::now();
+        attractive += std::chrono::duration<double, std::milli>(kernel6 - kernel5).count();
+
+        status = integrationKernelImpl<IdxType, DataType, cpu>(eta, momentum, exaggeration, posx, posy, attrx, attry, repx, repy, gainx, gainy,
+                                                               oldForcex, oldForcey, gradNorm, zNorm, nNodes, N);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        auto kernel7 = std::chrono::high_resolution_clock::now();
+        integration += std::chrono::duration<double, std::milli>(kernel7 - kernel6).count();
+        if ((i + 1) % nIterCheck == 0)
+        {
+            if (divergence < bestDivergence)
+            {
+                bestDivergence = divergence;
+                bestIter       = i;
+            }
+
+            if (gradNorm <= minGradNorm)
+            {
+                curIter = i;
+                break;
+            }
+            curIter = i;
+        }
+    }
+
+    momentum     = 0.8;
+    exaggeration = 1.;
+
+    for (IdxType i = explorationIter; i < maxIter; ++i)
+    {
```

Comment:
I seem to see duplicate code. Can it take it out as a separate function or change momentum and exaggeration in first FOR?